nupic/research/frameworks/pytorch/imagenet/experiment_utils.py

```diff
@@ -212,33 +210,25 @@ def create_optimizer(model, optimizer_class, optimizer_args,
         If False, remove 'weight_decay' from batch norm parameters
         See https://arxiv.org/abs/1807.11205
 
+    :param bias_weight_decay:
+        Whether or not to apply weight decay to bias modules parameters.
+        If False, remove 'weight_decay' from bias parameters
+
     :return: Configured optimizer
     """
-    if batch_norm_weight_decay:
-        # No need to remove weight decay. Use same optimizer args for all parameters
-        model_params = model.parameters()
-    else:
-        # Group batch norm parameters
-        def group_by_batch_norm(module):
-            return isinstance(module, _BatchNorm)
-
-        sorted_modules = sorted(model.modules(), key=group_by_batch_norm)
-        grouped_parameters = {
-            k: list(itertools.chain.from_iterable(m.parameters(False) for m in g))
-            for k, g in itertools.groupby(sorted_modules, key=group_by_batch_norm)
-        }
-
-        model_params = []
-        for is_bn, params in grouped_parameters.items():
-            # Group model_params
-            group_args = copy.deepcopy(optimizer_args)
-            group_args.update(params=params)
-
-            # Remove 'weight_decay' from batch norm parameters
-            if is_bn:
-                group_args.update(weight_decay=0.0)
-
-            model_params.append(group_args)
+
+    group_decay, group_no_decay = [], []
+    for name, param in model.named_parameters():
+        if not param.requires_grad:
+            continue
+        if ".bn" in name and not batch_norm_weight_decay:
```

Comment:
That is a very specific pattern, ".bn" (notice the dot). We are also using a very specific network, which is under our control, there are no surprises here. I don't see why we should complicate what is simple.

Reply:
There is an issue here, I can't check module type. I only get name and the tensor using named parameters. It is either this or change to modules as it was before.

Reply:
If it is specific to the network then add a method to the network class and call that method instead of this function. From the specific network class you can freely access any of the network modules by name or directly.

The current version of this function is not specific to the network implementation, you can pass any ResNet implementation (pytorch or custom) and it will update the batchnorm modules using the module type without any problems, you can adapt the iterator logic to also update the bias.

Relying on naming convention can cause many unintended problems and it is really hard to maintain.

Reply:
Thanks for the snippet, that is helpful. Just to be clear, I am not introducing a dependency on naming - there are references elsewhere on that same file, such as `m.bn2.weight` or `m.regular_path`, which is why I didn't think it would go against the conventions to use naming in that function as well. In general, I agree with being generalizable as long as we are consistent throughout the codebase.

Reply:
In the examples you mentioned the code is checking the object instance type before accessing a member property just like you would do with any other python class.
Which is different than selecting parameters from a list of parameters assuming the parameter name matching partial string.

Reply:
For those non-generalizable cases such as `n.bn2.weight`, notice they're in a function called `init_resnet50_batch_norm` and are proceeded by checks such as `isinstance(m, torchvision.models.resnet.BasicBlock)`. Thus, they are actually serving a relatively specific purpose. One could argue that `create_optimizer` is intended to be more general.

Reply:
Implemented, following suggestion.